### ISSUE TITLE
Removing Elements/Nodes Asynchronously(?)

### ISSUE THREAD

**Issue by dylan** [hyperapp/hyperapp#310]
Here's a fiddle to illustrate my use case: https://jsfiddle.net/dylanrw/mcg324a4/

Typically there isn't a problem, if state changes, I want the irrelevant elements removed. But when it comes to animating elements before removal, it would be nice if I could begin a transition and have the removal/sorting process occur once that transition is completed, or do something to have the soon-to-be-removed element stay in place until the transition is complete (even if the other elements update). (As you see in the fiddle, if you click an element, the elements are sorted, and then the element clicked is removed at the bottom.)

I tried one experiment at @jbucaran's encouragement where I added a lifecycle event `onreorder`, and I noticed that I got the results I was after by skipping the `element.insertBefore` in [app.js#265](https://github.com/hyperapp/hyperapp/blob/master/src/app.js#L265). I wonder if something like a `noreorder` node attribute that does similar would work?

Thanks in advance to @jbucaran and @zaceno for their collective patience!

**Comment by zaceno** [comment]
I think the title of this issue is probably incorrect. Or at least the "element" part anyway. Because the problem here isnt that the element needs to stay in the dom until a transition is done (it already does) but once the keyed node is no longer in the virtual tree, the patch algorithm will shuffle the elements around, causing the element that will be removed later, to find a new position among its siblings.

I suppose it would be possible to introduce some kind of state that keeps the element in the virtual node tree until further notice (even if not in the tree that view returns) ... but, I don't see any obvious general soloutions.

I think I would either:
- Manage the delayed removal myself, in the app state. So keep it in the vnode tree until it's done transitioning out.
- Or, If I wanted a general and reusable solution, I would track the elements' positions like I do in my transitions lib. Although that is far from elegant, and probably quite inefficient (not to mention spewing data attributes all over your DOM)

 So while there are soloutions, none are optimal, and it's good of @dylan to bring it up!

**Comment by jbucaran** [comment]
Is https://github.com/hyperapp/hyperapp/pull/312 related to this?

**Comment by zaceno** [comment]
@jbucaran Not related, no. Or... only tangentially:

My second option for working around the issue @dylan brought up, which is to track element coordinates, requires notification when layout changes. #312 is a solution for that.

An alternative to #312 would be to just attach dummy props to the nodes, with function value or Math.random(), to ensure that onupdate gets called every state change.

**Comment by jbucaran** [comment]
@zaceno So what you want is an event that runs on DOM nodes every time the state changes?

**Comment by zaceno** [comment]
@jbucaran For the use case of tracking positions, every state-change *works*, but is *wasteful*. It would be enough with an event when layout changes, and #312 is trying to provide that.

**Comment by zaceno** [comment]
@dylan Here's a codepen where we leave it up to userland to delay the removal. Unfortunately I don't think a built in solution is realistic: 

https://codepen.io/zaceno/pen/WEeqMO?editors=0110

**Comment by dylan** [comment]
@zaceno Yeah I can definitely do this, but it isn't really conducive to a non-position-tracking-by-muddying-the-data-attributes-transform-group-esque component. I think I'll watch the hyperapp api and if a useful change comes along I'll attempt it again.

Thanks!

**Comment by zaceno** [comment]
@dylan I get that. Even if nothing comes along -- if you just have an idea of what that useful thing you're waiting for to come along might be, ping me! Because while I agree there is something here that could be made better, I have a hard time seeing how.

**Comment by dylan** [comment]
Do we have something like `componentWillReceiveProps`? I was looking to see if there's anything to learn from react and that looks handy...

**Comment by jbucaran** [comment]
@dylan Sorry, I've been to busy to follow this up. Why did you close here?

**Comment by dylan** [comment]
It seems there's a disconnect between my description/desires and how Hyper works and I didn't want to waste everyone's time with dead ends. :P

Feel free to reopen it if you'd like.

**Comment by jbucaran** [comment]
@dylan Sorry, like I said I have been busy and didn't have time to follow up this issue.

You can't expect issues to resolved just in a few days. FYI sometimes issues remain open for longer than 3 days before anything actionable can even happen. Also, it's not up to you to decide if you are wasting my time or not. 

AFIC you were not wasting my time here.

TBH I was planning to have a look in a few days or maybe in a week or two.

TLDR If you can't help yourself and no one else can help much before I get a chance to get to you, then just leave things as they are and make sure you provide all the information you can to help us help you when the time is right. Even if you are not available or even interested when that happens.

**Comment by dylan** [comment]
> You can't expect issues to be resolved in a couple of days. FYI, issues are sometimes open for longer than 3 days before anything actionable can happen.

Hopefully you didn't consider my "waste of time" comment sarcasm or a snide remark about responsiveness, my initial writeup/problem/goals are just fuzzy enough that I was self conscious about my taking up you guys' time.

In the future I won't close any issues.

> P.S: Also, it's not up to you to decide if you are wasting my time or not. AFIC you are/were not wasting my time. 😉

Cool, reopen the issue if you want.

Slightly back on topic however, something like react's `componentWillReceiveProps` may allow me to pull off what I was originally after, but I need to rethink how I'm approaching the problem with Hyper to better articulate things.

Thanks again!

**Comment by jbucaran** [comment]
`componentWillReceiveProps` is very similar to what @zaceno and I were talking about, `onpatch`. 🤔
